File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const stack = [];
    const mappings = {
        ')':'(',
        '}':'{',
        ']':'['
    };

    for (let char of s) {
        // Check if character is a closing bracket
        if (mappings[char]) {
            // Pop the top element from stack. If the stack is empty, assign a dummy value '#'
            const topElement = stack.pop() || '#';

            // Check the popped element against the current character
            if (topElement !== mappings[char]) {
                return false;
            }
        } else {
            // It's opening bracket, push onto the stack
            stack.push(char);
        }
    }

    // If the stack still has elements, then it's unbalanced
    return stack.length === 0;
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        # Map to hold matching pairs of brackets
        bracket_map = {')': '(', '}': '{', ']': '['}
        stack = []
        
        for char in s:
            if char in bracket_map:  # If it's a closing bracket
                top_element = stack.pop() if stack else '#'
                if bracket_map[char] != top_element:
                    return False
            else:  # If it's an opening bracket
                stack.append(char)
        
        # If the stack is empty, all brackets were properly closed and nested
        return not stack
